pages/design_documentation/ocf_io_engines.md

```diff
@@ -0,0 +1,292 @@
+---
+title: Cache modes implementation - I/O engines.
+last_updated: August 17, 2022
+toc: false
+permalink: ocf_io_engines.html
+---
+
+Open CAS implements several [**cache modes**](/cache_configuration.html), 
+each suitable for a different usage scenario. To achieve a desired mode of 
+operation, Open CAS Framework uses basic building blocks called **IO engines**. 
+These blocks implement specific algorithms, e.g. write-through or write-back 
+and handle special or corner cases as well.
+
+
+# OCF I/O engines - general description.
+In general, OCF I/O engines are basic building blocks of a logic behind a 
+particular cache mode. In other words, engines are sets of functions implementing 
+the logic behind write and read operations for a given cache mode 
+(and more, see below).
+
+This is an excerpt of a mapping table used in the OCF to call read or write
+handlers of a particular cache mode (*write-through* and *write-back* modes 
+mappings are shown). 
+
+~~~~~~~
+static const struct ocf_io_if IO_IFS[OCF_IO_PRIV_MAX_IF] = {
+	[OCF_IO_WT_IF] = {
+		.read = ocf_read_generic,       
+                .write = ocf_write_wt,          
+                .name = "Write Through"
+	}
+        [OCF_IO_WB_IF] = { 
+                .read = ocf_read_generic,       
+                .write = ocf_write_wb,          
+                .name = "Write Back"            
+        },
+[...]
+~~~~~~~
+
+For these particular engines, only the write path engines are different and
+specific engine functions are assigned accordingly. The read path is handled 
+by a generic read function.
+
+## I/O request handling by the engines
+
+When an I/O request is issued from an application using OpenCAS, the framework
+determines which engine to use to handle this request and then 
+a proper handling function is called. 
+
+Most of the time, the engine implementing a configured cache mode read or write
+path is used. For example, for a write request to a cache running in a *write-back*
+mode - *ocf_write_wb()* will be called.
+
+However, in some less usual cases, a different I/O engine may be used 
+regardless of the configured cache mode (e.g. sequential cutoff threshold hit,
+cache drive I/O error, unaligned I/O, etc.).
+
+
+# OCF I/O engines - detailed description.
+Open CAS Framework uses the following engines:
```

Comment:
I would add in brackets the short version of the engine names used later for clarity and easier searching.